torch_geometric/distributed/dist_neighbor_loader.py

```diff
@@ -0,0 +1,154 @@
+            transform: Optional[Callable] = None,
+            transform_sampler_output: Optional[Callable] = None,
+            is_sorted: bool = False,
+            directed: bool = True,  # Deprecated.
+            with_edge: bool = False,
+            concurrency: int = 1,
+            collect_features: bool = True,
+            filter_per_worker: Optional[bool] = False,
+            async_sampling: bool = True,
+            device: torch.device = torch.device('cpu'),
+            **kwargs,
+    ):
+
+        assert (isinstance(data[0], FeatureStore) and (data[1], GraphStore)
+                ), "Data needs to be Tuple[LocalFeatureStore, LocalGraphStore]"
+
+        assert concurrency >= 0, "concurrency must be greater than 1"
+
+        if input_time is not None and time_attr is None:
+            raise ValueError("Received conflicting 'input_time' and "
+                             "'time_attr' arguments: 'input_time' is set "
+                             "while 'time_attr' is not set.")
+
+        channel = torch.multiprocessing.Queue() if async_sampling else None
+
+        if neighbor_sampler is None:
+            neighbor_sampler = DistNeighborSampler(
+                data=data, current_ctx=current_ctx,
+                rpc_worker_names=rpc_worker_names, num_neighbors=num_neighbors,
+                with_edge=with_edge, replace=replace,
+                subgraph_type=subgraph_type, disjoint=disjoint,
+                temporal_strategy=temporal_strategy, time_attr=time_attr,
+                is_sorted=is_sorted, share_memory=kwargs.get('num_workers', 0)
+                > 0, directed=directed, device=device, channel=channel,
+                concurrency=concurrency, collect_features=collect_features)
+
+        DistLoader.__init__(self, neighbor_sampler=neighbor_sampler,
+                            channel=channel, master_addr=master_addr,
+                            master_port=master_port, current_ctx=current_ctx,
+                            rpc_worker_names=rpc_worker_names, **kwargs)
+        NodeLoader.__init__(self, data=data, node_sampler=neighbor_sampler,
+                            input_nodes=input_nodes, input_time=input_time,
+                            transform=transform,
+                            transform_sampler_output=transform_sampler_output,
+                            filter_per_worker=filter_per_worker,
+                            worker_init_fn=self.worker_init_fn, **kwargs)
+
+    def filter_fn(
+        self,
+        out: Union[SamplerOutput, HeteroSamplerOutput],
+    ) -> Union[Data, HeteroData]:
+        r"""Joins the sampled nodes with their corresponding features,
+        returning the resulting :class:`~torch_geometric.data.Data` or
+        :class:`~torch_geometric.data.HeteroData` object to be used downstream.
+        """
+        # TODO: Align dist_sampler metadata output with original pyg sampler, such that filter_fn() from the NodeLoader can be used
```

Comment:
What currently prevents us from using `filter_fn` from `NodeLoader`? I know we talked about this but can you refresh my knowledge?

Reply:
We had to align DistSampler output with PyG format. Now I've managed to use the standard `filter_fn` in both `NodeLoader` and `LinkLoader`. To facilitate this for hetero I needed to add an extra method in `loader.utils`. Please let me know if this solution is acceptable.